src/node.rs

```diff
@@ -348,29 +418,45 @@ impl<'a> NodeExt<'a> for Node<'a> {
     fn children_of_kind(&self, kind: Kind) -> ChildrenOfKind<'a> {
         ChildrenOfKind::new(*self, kind)
     }
+
+    fn parent_(&self, node_parent_provider: &impl NodeParentProvider<'a>) -> Node<'a> {
+        node_parent_provider.node_parent(*self)
+    }
+
+    fn maybe_parent(&self, node_parent_provider: &impl NodeParentProvider<'a>) -> Option<Node<'a>> {
+        node_parent_provider.maybe_node_parent(*self)
+    }
 }
 
-fn walk_cursor_to_descendant(cursor: &mut TreeCursor, node: Node) {
+fn walk_cursor_to_descendant<'a>(cursor: &mut TreeCursor<'a>, node: Node<'a>) {
     while cursor.node() != node {
         // this seems like it should be right but see https://github.com/tree-sitter/tree-sitter/issues/2463
         // cursor.goto_first_child_for_byte(node.start_byte()).unwrap();
-        if node.is_descendant_of(cursor.node()) {
+        fn is_inside(node: Node, cursor_node: Node) -> bool {
```

Comment:
Here I avoided having the need for the additional `node_parent_provider` parameter (which seemed at the time like it was then "infecting" a bunch of stuff but I think it might've ended up getting added anyway) because I think we don't really need to do true "parent" lookups for this algorithm